src/modules/emote_autocomplete/twitch/EmoteAutocomplete.jsx

```diff
@@ -0,0 +1,182 @@
+    return null;
+  }
+  try {
+    const [emoteProvider, emoteId, emoteCode] = atob(decodeURIComponent(emoteData)).split('-');
+    return {
+      provider: emoteProvider,
+      id: emoteId,
+      code: emoteCode,
+    };
+  } catch (_) {
+    return null;
+  }
+}
+
+function getAutocompleteEmoteProvider() {
+  const autocompleteNode = twitch.getAutocompleteStateNode();
+  if (autocompleteNode == null) {
+    return null;
+  }
+
+  const autocompleteProviders = autocompleteNode.stateNode.providers;
+  return autocompleteProviders.find(({autocompleteType}) => autocompleteType === 'emote');
+}
+
+function createTwitchEmoteSet(allEmotes) {
+  return {
+    emotes: allEmotes.map((emote) => ({
+      id: serializeEmoteId(emote),
+      modifiers: null,
+      setID: CUSTOM_SET_ID,
+      token: emote.code,
+    })),
+    id: CUSTOM_SET_ID,
+  };
+}
+
+async function injectEmoteSets() {
+  const autocompleteEmoteProvider = getAutocompleteEmoteProvider();
+  if (autocompleteEmoteProvider == null) {
+    return;
+  }
+
+  const autocompleteEmotes = autocompleteEmoteProvider.props.emotes;
+
+  const allEmotes = emotes.getEmotesByCategories([
+    EmoteCategories.BETTERTTV_CHANNEL,
+    EmoteCategories.BETTERTTV_GLOBAL,
+    EmoteCategories.BETTERTTV_PERSONAL,
+    EmoteCategories.FRANKERFACEZ_CHANNEL,
+    EmoteCategories.FRANKERFACEZ_GLOBAL,
+  ]);
+
+  const emoteSet = createTwitchEmoteSet(allEmotes);
+  const index = autocompleteEmotes.findIndex(({id}) => id === CUSTOM_SET_ID);
+
+  if (index === -1) {
+    autocompleteEmotes.push(emoteSet);
+  } else {
+    autocompleteEmotes[index] = emoteSet;
+  }
+
+  autocompleteEmoteProvider.forceUpdate();
+}
+
+let cleanup = null;
+export default class EmoteAutocomplete {
+  constructor() {
+    watcher.on('channel.updated', () => this.load());
+    settings.on(`changed.${SettingIds.EMOTE_AUTOCOMPLETE}`, () => this.load());
+  }
+
+  // on chat load, add emotes to twitch's autocomplete emote provider
+  // on autocomplete match render, replace image source of betterttv emotes with bttv cdn links
+
+  patchEmoteImage(image, isConnected) {
+    if (!isConnected) {
+      return;
+    }
+
+    const url = String(image.srcset);
+    const deseralizedEmote = deserializeEmoteFromURL(url);
+
+    if (deseralizedEmote == null) {
+      return;
+    }
+
+    const emote = emotes.getEligibleEmote(deseralizedEmote.code);
+    if (emote == null) {
+      return;
+    }
+
+    image.srcset = createSrcSet(emote.images);
+    image.src = emote.images['1x'];
+  }
+
+  unload() {
+    cleanup();
+    cleanup = null;
+
+    const autocompleteEmoteProvider = getAutocompleteEmoteProvider();
+    if (autocompleteEmoteProvider == null) {
+      return;
+    }
+
+    autocompleteEmoteProvider.props.emotes = autocompleteEmoteProvider.props.emotes.filter(
+      ({id}) => id !== CUSTOM_SET_ID
+    );
+
+    autocompleteEmoteProvider.forceUpdate();
+  }
+
+  load() {
+    if (cleanup != null) {
+      this.unload();
```

Comment:
why do we unlisten just to end up listening again? is there a more optimized way to write this logic?